power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp

```diff
@@ -0,0 +1,471 @@
+            return -1;
+        }
+        if (batch_size_ == 0) {
+            return 0;
+        }
+        return counter.front();
+    }
+
+    void parse_component(Buffer const& buffer) const {
+        component_key_ = buffer.component->name;
+        // handle indptr
+        if (buffer.elements_per_scenario < 0) {
+            // first always zero
+            buffer.indptr.front() = 0;
+            // accumulate sum
+            // TODO Apple Clang cannot compile transform_inclusive_scan correctly
+            // So we disable the good code and write the loop manually
+            // std::transform_inclusive_scan(buffer.msg_data.cbegin(), buffer.msg_data.cend(),
+            //                               buffer.indptr.begin() + 1,
+            //                               std::plus{}, [](auto const& x) { return static_cast<Idx>(x.size()); });
+            for (Idx i = 0; i != batch_size_; ++i) {
+                buffer.indptr[i + 1] = buffer.indptr[i] + static_cast<Idx>(buffer.msg_data[i].size());
+            }
+        }
+        // set nan
+        buffer.component->set_nan(buffer.data, 0, buffer.total_elements);
+        // attributes
+        auto const attributes = [&]() -> std::span<MetaAttribute const* const> {
+            auto const found = attributes_.find(buffer.component->name);
+            if (found == attributes_.cend()) {
+                return {};
+            }
+            return found->second;
+        }();
+        // all scenarios
+        for (scenario_number_ = 0; scenario_number_ != batch_size_; ++scenario_number_) {
+            Idx const scenario_offset = buffer.elements_per_scenario < 0
+                                            ? buffer.indptr[scenario_number_]
+                                            : scenario_number_ * buffer.elements_per_scenario;
+#ifndef NDEBUG
+            if (buffer.elements_per_scenario < 0) {
+                assert(buffer.indptr[scenario_number_ + 1] - buffer.indptr[scenario_number_] ==
+                       static_cast<Idx>(buffer.msg_data[scenario_number_].size()));
+
+            } else {
+                assert(buffer.elements_per_scenario == static_cast<Idx>(buffer.msg_data[scenario_number_].size()));
+            }
+#endif
+            void* scenario_pointer = reinterpret_cast<char*>(buffer.data) + scenario_offset * buffer.component->size;
+            parse_scenario(*buffer.component, scenario_pointer, buffer.msg_data[scenario_number_], attributes);
+        }
+        scenario_number_ = -1;
+        component_key_ = "";
+    }
+
+    void parse_scenario(MetaComponent const& component, void* scenario_pointer, ArraySpan msg_data,
+                        std::span<MetaAttribute const* const> attributes) const {
+        for (element_number_ = 0; element_number_ != static_cast<Idx>(msg_data.size()); ++element_number_) {
+            void* element_pointer = reinterpret_cast<char*>(scenario_pointer) + element_number_ * component.size;
```

Comment:
can you move this line to another function? i would prefer to isolate `reinterpret_cast`s as much as possible

Reply:
done